include/swift/AST/AccessScope.h

```diff
@@ -48,7 +53,7 @@ class AccessScope {
   /// \see DeclContext::isChildContextOf
   bool isChildOf(AccessScope AS) const {
     if (!isPublic() && !AS.isPublic())
-      return getDeclContext()->isChildContextOf(AS.getDeclContext());
+      return allowsPrivateAccess(getDeclContext(), AS.getDeclContext());
```

Comment:
(Ah, I see now that NameLookup.cpp tries to access this logic directly. It might just make sense to change that `checkAccessibility` over to using access scopes explicitly instead of trying to be clever and do everything with DeclContexts.)